AutoGraph/WebSocketClient.swift

```diff
@@ -0,0 +1,287 @@
+        DispatchQueue.main.asyncAfter(deadline: .now() + delayInSeconds) { [weak self] in
+            guard let self = self else { return }
+            // Requeue all so they don't get error callbacks on disconnect and they get re-subscribed on connect.
+            self.requeueAllSubscribers()
+            self.disconnect()
+            self.webSocket.connect()
+        }
+        return delayInSeconds
+    }
+    
+    // TODO: needs a good test
+    /// Takes all subscriptions and puts them back on the queue, used for reconnection.
+    func requeueAllSubscribers() {
+        for (_, subscriptionSet) in self.subscriptions {
+            for (subscriber, subscriptionResponseHandler) in subscriptionSet {
+                self.queuedSubscriptions[subscriber] = connectionCompletionBlock(subscriber: subscriber, responseHandler: subscriptionResponseHandler)
+            }
+        }
+        self.subscriptions.removeAll(keepingCapacity: true)
+    }
+    
+    // TODO: test
+    /// Take all connection completion blocks out of the queue and runs them. Removes from queue first to avoid any side affects.
+    func didConnect() throws {
+        self.state = .connected
+        self.attemptReconnectCount = kAttemptReconnectCount
+        self.reconnecting = false
+        
+        let connectedPayload = try GraphQLWSProtocol.connectionInit.serializedSubscriptionPayload()
+        self.write(connectedPayload)
+        
+        let queuedSubscriptions = self.queuedSubscriptions
+        self.queuedSubscriptions.removeAll()
+        queuedSubscriptions.forEach { (_, connected: WebSocketConnected) in
+            connected(.success(()))
+        }
+    }
+    
+    func reset() {
+        self.disconnect()
+        self.subscriptions.removeAll()
+        self.queuedSubscriptions.removeAll()
+        self.attemptReconnectCount = kAttemptReconnectCount
+        self.reconnecting = false
+    }
+}
+
+// MARK: - Class Method
+
+extension WebSocketClient {
+    class func connectionRequest(url: URL) throws -> URLRequest {
+        var defaultHeders = [String: String]()
+        defaultHeders["Sec-WebSocket-Protocol"] = "graphql-ws"
+        defaultHeders["Origin"] = url.absoluteString
+        
+        return try URLRequest(url: url, method: .get, headers: HTTPHeaders(defaultHeders))
+    }
+}
+
+// MARK: - WebSocketDelegate
+
+extension WebSocketClient: WebSocketDelegate {
+    // This is called on the Starscream callback queue, which defaults to Main.
+    public func didReceive(event: WebSocketEvent, client: WebSocket) {
+        self.delegate?.didReceive(event: event)
+        
+        // If we get any event at all just assume that the previous reconnect attempt either succeeded or failed for simplicity.
+        self.reconnecting = false
+        
+        do {
+            switch event {
+            case .disconnected, .cancelled:
+                _ = self.reconnect()
```

Comment:
Actually thinking about this. If the user manually disconnects the WebSocket. That will send the even .disconnected and will attempt to still reconnect. Maybe we should wrap a flag around it if user manually disconnect else then reconnect

Reply:
they can just destroy the whole autograph and build a new one if that's their M.O.

Reply:
something we might want to add later though

Reply:
okay

Reply:
we also might want to change the name of `disconnect` to `disconnectAndReconnect` then too for now